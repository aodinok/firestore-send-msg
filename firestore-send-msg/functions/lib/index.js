"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const admin = __importStar(require("firebase-admin"));
const functions = __importStar(require("firebase-functions"));
const messagebird_1 = __importDefault(require("messagebird"));
const config_1 = __importDefault(require("./config"));
const log_1 = require("./log");
let db;
let mb;
let initialized = false;
/**
 * Initializes Admin SDK & MessageBird client if not already initialized.
 */
function initialize() {
    if (initialized === true)
        return;
    initialized = true;
    log_1.logInfo('initializing app...');
    admin.initializeApp();
    log_1.logInfo('initializing db...');
    db = admin.firestore();
    log_1.logInfo('initializing mb api client...');
    mb = messagebird_1.default(config_1.default.accessKey);
    log_1.logInfo('initialization finished successfuly');
}
function deliver(payload, ref) {
    return __awaiter(this, void 0, void 0, function* () {
        const update = {
            "delivery.attempts": admin.firestore.FieldValue.increment(1),
            "delivery.endTime": admin.firestore.FieldValue.serverTimestamp(),
            "delivery.error": null,
            "delivery.leaseExpireTime": null,
        };
        log_1.logInfo('delivery attempt');
        try {
            if (!payload.channelId) {
                throw new Error("Failed to deliver message. ChannelId is not defined.");
            }
            if (!payload.to) {
                throw new Error("Failed to deliver message. Recipient of the message should be filled.");
            }
            if (!payload.content) {
                throw new Error("Failed to deliver message. Message content is empty.");
            }
            log_1.logInfo(`sending message to channelId: ${payload.channelId}`);
            log_1.logInfo(`with content: ${payload.content}`);
            yield yield new Promise((resolve, reject) => {
                mb.conversations.start(payload, function (err, response) {
                    if (err) {
                        log_1.logWarn(`send failed, got error: ${err}`);
                        return reject(err);
                    }
                    log_1.logInfo(`send successfully scheduled, got response: ${response}`);
                    update["messageId"] = response.id;
                    update["delivery.state"] = "SUCCESS";
                    resolve();
                });
            });
        }
        catch (e) {
            log_1.logInfo(`updating delivery record with error message`);
            update["delivery.state"] = "ERROR";
            update["delivery.error"] = e.toString();
        }
        return db.runTransaction((transaction) => {
            transaction.update(ref, update);
            return Promise.resolve();
        });
    });
}
function processCreate(snap) {
    return __awaiter(this, void 0, void 0, function* () {
        log_1.logInfo('new msg added, init delivery object for it');
        return db.runTransaction((transaction) => {
            transaction.update(snap.ref, {
                delivery: {
                    startTime: admin.firestore.FieldValue.serverTimestamp(),
                    state: "PENDING",
                    attempts: 0,
                    error: null,
                },
            });
            return Promise.resolve();
        });
    });
}
function processWrite(change) {
    return __awaiter(this, void 0, void 0, function* () {
        log_1.logInfo('processing write');
        if (!change.after.exists) {
            log_1.logInfo('ignoring delete');
            return null;
        }
        if (!change.before.exists && change.after.exists) {
            log_1.logInfo('process create');
            return processCreate(change.after);
        }
        const payload = change.after.data();
        log_1.logInfo('processing update');
        switch (payload.delivery.state) {
            case "SUCCESS":
            case "ERROR":
                log_1.logInfo('current state is SUCCESS/ERROR');
                return null;
            case "PROCESSING":
                log_1.logInfo('current state is PROCESSING');
                if (payload.delivery.leaseExpireTime.toMillis() < Date.now()) {
                    return db.runTransaction((transaction) => {
                        transaction.update(change.after.ref, {
                            "delivery.state": "ERROR",
                            error: "Message processing lease expired.",
                        });
                        return Promise.resolve();
                    });
                }
                return null;
            case "PENDING":
            case "RETRY":
                log_1.logInfo('current state is PENDING/RETRY');
                yield db.runTransaction((transaction) => {
                    transaction.update(change.after.ref, {
                        "delivery.state": "PROCESSING",
                        "delivery.leaseExpireTime": admin.firestore.Timestamp.fromMillis(Date.now() + 60000),
                    });
                    return Promise.resolve();
                });
                log_1.logInfo('record set to PROCESSING state, trying to deliver the message');
                return deliver(payload, change.after.ref);
        }
    });
}
exports.processQueue = functions.handler.firestore.document.onWrite((change) => __awaiter(void 0, void 0, void 0, function* () {
    initialize();
    try {
        yield processWrite(change);
    }
    catch (err) {
        log_1.logWarn('unexpected error during execution: ', err);
        return null;
    }
}));
