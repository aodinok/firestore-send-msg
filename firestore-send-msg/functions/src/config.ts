export default {
  location: process.env.LOCATION,
  accessKey: process.env.ACCESS_KEY,
  msgCollection: process.env.MSG_COLLECTION,    
  defaultOriginator: process.env.DEFAULT_ORIGINATOR
};