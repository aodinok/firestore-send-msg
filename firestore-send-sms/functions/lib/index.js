"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var admin = __importStar(require("firebase-admin"));
var functions = __importStar(require("firebase-functions"));
var messagebird_1 = __importDefault(require("messagebird"));
var config_1 = __importDefault(require("./config"));
var log_1 = require("./log");
var db;
var mb;
var initialized = false;
/**
 * Initializes Admin SDK & MessageBird client if not already initialized.
 */
function initialize() {
    if (initialized === true)
        return;
    initialized = true;
    log_1.logInfo("initializing app...");
    admin.initializeApp();
    log_1.logInfo("initializing db...");
    db = admin.firestore();
    log_1.logInfo("initializing messagebird client...");
    mb = messagebird_1.default(config_1.default.accessKey);
    log_1.logInfo("initialization finished successfuly");
}
function deliver(payload, ref) {
    return __awaiter(this, void 0, void 0, function () {
        var update, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    update = {
                        "delivery.attempts": admin.firestore.FieldValue.increment(1),
                        "delivery.endTime": admin.firestore.FieldValue.serverTimestamp(),
                        "delivery.error": null,
                        "delivery.leaseExpireTime": null,
                    };
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    if (!payload.recipients.length) {
                        throw new Error("Failed to deliver sms. Expected at least 1 recipient.");
                    }
                    payload.originator = payload.originator || config_1.default.defaultOriginator;
                    return [4 /*yield*/, new Promise(function (resolve, reject) {
                            mb.messages.create(payload, function (err, response) {
                                if (err) {
                                    return reject(err);
                                }
                                // TODO: update delivery state when message delivered or delivery failed, we may need to use status URL for this
                                log_1.logInfo("send successfully scheduled, got response: ", response);
                                update["messageId"] = response.id;
                                update["delivery.state"] = "SUCCESS";
                                resolve();
                            });
                        })];
                case 2:
                    _a.sent();
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _a.sent();
                    log_1.logInfo("updating delivery record with error message");
                    update["delivery.state"] = "ERROR";
                    update["delivery.error"] = e_1.toString();
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/, admin.firestore().runTransaction(function (transaction) {
                        transaction.update(ref, update);
                        return Promise.resolve();
                    })];
            }
        });
    });
}
function processCreate(snap) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            log_1.logInfo("new msg added, init delivery object for it");
            return [2 /*return*/, admin.firestore().runTransaction(function (transaction) {
                    transaction.update(snap.ref, {
                        delivery: {
                            startTime: admin.firestore.FieldValue.serverTimestamp(),
                            state: "PENDING",
                            attempts: 0,
                            error: null,
                        },
                    });
                    return Promise.resolve();
                })];
        });
    });
}
function processWrite(change) {
    return __awaiter(this, void 0, void 0, function () {
        var payload, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    log_1.logInfo("processing write");
                    if (!change.after.exists) {
                        log_1.logInfo("ignoring delete");
                        return [2 /*return*/, null];
                    }
                    if (!change.before.exists && change.after.exists) {
                        log_1.logInfo("process create");
                        return [2 /*return*/, processCreate(change.after)];
                    }
                    log_1.logInfo("processing update");
                    payload = change.after.data();
                    _a = payload.delivery.state;
                    switch (_a) {
                        case "SUCCESS": return [3 /*break*/, 1];
                        case "ERROR": return [3 /*break*/, 1];
                        case "PROCESSING": return [3 /*break*/, 2];
                        case "PENDING": return [3 /*break*/, 3];
                        case "RETRY": return [3 /*break*/, 3];
                    }
                    return [3 /*break*/, 5];
                case 1:
                    log_1.logInfo("current state is SUCCESS/ERROR");
                    return [2 /*return*/, null];
                case 2:
                    log_1.logInfo("current state is PROCESSING");
                    if (payload.delivery.leaseExpireTime &&
                        payload.delivery.leaseExpireTime.toMillis() < Date.now()) {
                        return [2 /*return*/, admin.firestore().runTransaction(function (transaction) {
                                transaction.update(change.after.ref, {
                                    "delivery.state": "ERROR",
                                    error: "Message processing lease expired.",
                                });
                                return Promise.resolve();
                            })];
                    }
                    return [2 /*return*/, null];
                case 3:
                    log_1.logInfo("current state is PENDING/RETRY");
                    return [4 /*yield*/, admin.firestore().runTransaction(function (transaction) {
                            transaction.update(change.after.ref, {
                                "delivery.state": "PROCESSING",
                                "delivery.leaseExpireTime": admin.firestore.Timestamp.fromMillis(Date.now() + 60000),
                            });
                            return Promise.resolve();
                        })];
                case 4:
                    _b.sent();
                    log_1.logInfo("record set to PROCESSING state, trying to deliver the message");
                    return [2 /*return*/, deliver(payload, change.after.ref)];
                case 5: return [2 /*return*/];
            }
        });
    });
}
exports.processQueue = functions.handler.firestore.document.onWrite(function (change) { return __awaiter(void 0, void 0, void 0, function () {
    var err_1;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                initialize();
                _a.label = 1;
            case 1:
                _a.trys.push([1, 3, , 4]);
                return [4 /*yield*/, processWrite(change)];
            case 2:
                _a.sent();
                return [3 /*break*/, 4];
            case 3:
                err_1 = _a.sent();
                log_1.logWarn("unexpected error during execution: ", err_1);
                return [2 /*return*/, null];
            case 4: return [2 /*return*/];
        }
    });
}); });
