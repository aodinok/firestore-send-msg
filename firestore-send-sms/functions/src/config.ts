export default {
  location: process.env.LOCATION,
  accessKey: process.env.ACCESS_KEY,
  smsCollection: process.env.SMS_COLLECTION,
  defaultOriginator: process.env.DEFAULT_ORIGINATOR,
  logLevel: process.env.LOG_LEVEL,
};
