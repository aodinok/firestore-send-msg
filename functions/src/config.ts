export default {
  location: process.env.LOCATION,
  accessKey: process.env.ACCESS_KEY,
  msgCollection: process.env.MSG_COLLECTION,
  logLevel: process.env.LOG_LEVEL,
};
