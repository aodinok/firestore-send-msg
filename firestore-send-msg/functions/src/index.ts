import * as admin from "firebase-admin";
import * as functions from "firebase-functions";
import messagebird, { StartConversationParameter, MessageBird } from "messagebird";

import config from "./config";

// in https://github.com/messagebird/messagebird-nodejs/blob/master/types/messages.d.ts#L7 there is a better type definition
// we should try to make this plugin a very transparent proxy. Rsing the rest sdk type defs would be a step in that direction.
interface QueuePayload extends StartConversationParameter {
  messageId?: string;
  originator: string | number;
  body: string;
  recipients: string[];

  delivery?: {
    startTime: FirebaseFirestore.Timestamp;
    endTime: FirebaseFirestore.Timestamp;
    leaseExpireTime: FirebaseFirestore.Timestamp;
    state: "PENDING" | "PROCESSING" | "RETRY" | "SUCCESS" | "ERROR";
    attempts: number;
    error?: string;
  };
}

let db: FirebaseFirestore.Firestore;
let mb: MessageBird;
let initialized = false;

/**
 * Initializes Admin SDK & MessageBird client if not already initialized.
 */
function initialize() {
  if (initialized === true) return;
  initialized = true;
  admin.initializeApp();
  db = admin.firestore();
  mb = messagebird(config.accessKey);
}

async function deliver(
  payload: QueuePayload,
  ref: FirebaseFirestore.DocumentReference
): Promise<any> {
  const update = {
    "delivery.attempts": admin.firestore.FieldValue.increment(1),
    "delivery.endTime": admin.firestore.FieldValue.serverTimestamp(),
    "delivery.error": null,
    "delivery.leaseExpireTime": null,
  };

  try {
    if (!payload.recipients.length) {
      throw new Error("Failed to deliver sms. Expected at least 1 recipient.");
    }

    payload.originator = payload.originator || config.defaultOriginator;

    // TODO: channel and content should come from payload
    await mb.conversations.start({
      'to': payload.recipients[0],
      'channelId': '6730dba0444b46d7976d44b57a8bb9e3' ,
      'type': 'text',
      'content': { 'text': 'Hello From Firebase extension!'  }
    }, function (err, response) {
      if (err) {
        throw err;
      }
      console.log('got response: ', response);
      update["messageId"] = response.id;
    });

  } catch (e) {
    update["delivery.state"] = "ERROR";
    update["delivery.error"] = e.toString();
  }

  return admin.firestore().runTransaction((transaction) => {
    transaction.update(ref, update);
    return Promise.resolve();
  });
}

async function processCreate(snap: FirebaseFirestore.DocumentSnapshot) {
  return admin.firestore().runTransaction((transaction) => {
    transaction.update(snap.ref, {
      delivery: {
        startTime: admin.firestore.FieldValue.serverTimestamp(),
        state: "PENDING",
        attempts: 0,
        error: null,
      },
    });
    return Promise.resolve();
  });
}

async function processWrite(change) {
  if (!change.after.exists) {
    return null;
  }

  if (!change.before.exists && change.after.exists) {
    return processCreate(change.after);
  }

  const payload = change.after.data() as QueuePayload;

  switch (payload.delivery.state) {
    case "SUCCESS":
    case "ERROR":
      return null;
    case "PROCESSING":
      if (payload.delivery.leaseExpireTime.toMillis() < Date.now()) {
        return admin.firestore().runTransaction((transaction) => {
          transaction.update(change.after.ref, {
            "delivery.state": "ERROR",
            error: "Message processing lease expired.",
          });
          return Promise.resolve();
        });
      }
      return null;
    case "PENDING":
    case "RETRY":
      await admin.firestore().runTransaction((transaction) => {
        transaction.update(change.after.ref, {
          "delivery.state": "PROCESSING",
          "delivery.leaseExpireTime": admin.firestore.Timestamp.fromMillis(
            Date.now() + 60000
          ),
        });
        return Promise.resolve();
      });
      return deliver(payload, change.after.ref);
  }
}

export const processQueue = functions.handler.firestore.document.onWrite(
  async (change) => {
    initialize();
    try {
      await processWrite(change);
    } catch (err) {
      return null;
    }
  }
);
