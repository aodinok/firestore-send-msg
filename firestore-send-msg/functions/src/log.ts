import config from "./config";

export function logInfo(msg: string) {
  if (msg && config.logLevel === 'info') {
    console.info(`INFO: firestore-send-msg: ${msg}`)
  }
}

export function logWarn(msg: string) {
  if (msg && (config.logLevel === 'warn' || config.logLevel === 'info')) {
    console.warn(`WARN: firestore-send-msg: ${msg}`)
  }
}