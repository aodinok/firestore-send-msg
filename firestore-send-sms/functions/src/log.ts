import config from "./config";

export function logInfo(msg: string, ...rest) {
  if (msg && config.logLevel === "info") {
    console.info(`INFO: firestore-send-sms: ${msg}`, ...rest);
  }
}

export function logWarn(msg: string, ...rest) {
  if (msg && (config.logLevel === "warn" || config.logLevel === "info")) {
    console.warn(`WARN: firestore-send-sms: ${msg}`, ...rest);
  }
}
