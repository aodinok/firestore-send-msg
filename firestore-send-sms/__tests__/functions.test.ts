
import * as firebase from '@firebase/testing';
import functionsConfig from "../functions/src/config";
import * as exportedFunctions from "../functions/src";

const projectId = 'conversations-example'
const app = firebase.initializeTestApp({
  projectId,
  auth: { uid: "alice", email: "alice@example.com" }
});
const db = app.firestore();

describe("firestore-send-sms", () => {
  beforeEach(async () => {
    // Clean database before each test
    await firebase.clearFirestoreData({ projectId });
  });

  afterAll(async () => {
    await Promise.all(firebase.apps().map(app => app.delete()));
  });

  test("functions configuration detected from environment variables", async () => {
    expect(functionsConfig).toMatchSnapshot();
  });

  test("functions are exported", () => {
    expect(exportedFunctions.processQueue).toBeInstanceOf(Function);
  });

  // you need to have emulator runnning for this test
  test.skip("add to sms collection triggers processQueue function", async () => {
    // add new message to collection
    await db.collection('sms').add({
      originator: 'FunFacts',
      body: 'Fun Fact of the day: Dolly Parton lost in a Dolly Parton look-alike contest',
      recipients: [
        "+31617569806",
      ],
    });

    // wait for extension to trigger
    await new Promise((r) => setTimeout(r, 2000));

    // should update delivery with error state
    let allMsgs = await db.collection('sms').get();
    for(const doc of allMsgs.docs) {
      const data = doc.data()
      expect(data).toHaveProperty('delivery');
      expect(data.delivery).toHaveProperty('state');
      expect(data.delivery.state).toEqual('ERROR');
    }
  });
});
