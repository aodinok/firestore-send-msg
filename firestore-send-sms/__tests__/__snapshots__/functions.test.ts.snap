// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`firestore-send-sms functions configuration detected from environment variables 1`] = `
Object {
  "accessKey": "aaa",
  "defaultOriginator": "Marcel",
  "location": "europe-west1",
  "logLevel": "info",
  "smsCollection": "sms",
}
`;
